### PR TITLE
Add animations to blog website

### PR DESCRIPTION
**Pull request by its-kunal**
This PR fixes the issue - [Link](https://github.com/keploy/keploy/issues/1664) 

- [x] Read Progress Scrollbar
- [x] Hover effects for table of content items
- [x] Add Slide In Fade In animation for the preview cards.

### CODE REVIEW

**Comment by nehagup** [comment]
> Framer motion library

@its-kunal I haven't explored this library yet, however, if it's free to use (without limitations on time), we can use that.

**Comment by its-kunal** [comment]
@nehagup Sure, Do you have any suggestions for the changes?

**Review by Hermione2408 — CHANGES_REQUESTED** [review]
Hey @its-kunal Observed a few UI issues :
Screenshot from 2024-03-29 00-41-08

- Table of Contents (ToC) hover state causes text to overflow and unintentionally adds a scrollbar. You can contain the text within the div to prevent overflow and it should be aligned from left.
- Good work on the Progress bar but the progress bar should track the blog's content only, starting with the blog's introduction and ending before the 'More Stories' section. Currently, it spans the entire container including these elements.

**Comment by its-kunal** [comment]
Hey @Hermione2408, @nehagup please review my pull request, done the changes for -
- Read scroll progress bar, limited to the blog contents only.

Also, fade in animation for the post preview cards, and more posts by the author.

**Comment by nehagup** [comment]
@Hermione2408 please review.

**Comment by Hermione2408** [comment]
Hey @its-kunal build is failing right now please fix it 
Screenshot from 2024-04-01 10-41-53

**Comment by its-kunal** [comment]
Hey @Hermione2408, I have made the changes to fix the error.

File: pages/technology/[slug].tsx [hunk 1]
```diff
@@ -34,6 +36,24 @@ const postBody = ({ content, post }) => {
 export default function Post({ post, posts, preview }) {
   const router = useRouter();
   const morePosts = posts?.edges;
+  const postBodyRef = useRef<HTMLDivElement>();
+  const a = useSpringValue(0);
```
Comment (Hermione2408) [inline pages/technology/[slug].tsx:40]: use a descriptive name for the variable

File: pages/community/[slug].tsx [hunk 2]
```diff
@@ -35,6 +36,24 @@ const postBody = ({ content, post }) => {
 export default function Post({ post, posts, preview }) {
   const router = useRouter();
   const morePosts = posts?.edges;
+  const postBodyRef = useRef<HTMLDivElement>();
+  const a = useSpringValue(0);
```
Comment (Hermione2408) [inline pages/community/[slug].tsx:40]: here too

File: components/postByAuthorMapping.tsx [hunk 3]
```diff
@@ -1,43 +1,72 @@
...
+        easing: easings.easeInCubic,
+      },
+    }),
+    {
+      rootMargin: "-200px 0px",
+    }
+  );
+  return (
+    <animated.li className="mb-8" ref={ref} style={springStyles}>
+      <a href={`/blog/${node.categories.edges[0].node.name}/${node.slug}`}>
```
Comment (Hermione2408) [inline components/postByAuthorMapping.tsx:27]: since we are using next you can use <Link> instead of <a>

**Review by Hermione2408 — CHANGES_REQUESTED** [review]

File: components/postByAuthorMapping.tsx (hunk 3 above)
Comment (its-kunal) [reply components/postByAuthorMapping.tsx:27]: I'm sorry, I could not understand, can you please explain this change?

Comment (Hermione2408) [reply components/postByAuthorMapping.tsx:27]: you can use Link tag instead of anchor tag

File: pages/technology/[slug].tsx (hunk 1 above)
Comment (its-kunal) [reply pages/technology/[slug].tsx]: made the changes please refer to the commit- [817c2aa](https://github.com/keploy/blog-website/pull/31/commits/817c2aac2993b1939da6a7049aaf723305525bde)

File: pages/community/[slug].tsx (hunk 2 above)
Comment (its-kunal) [reply pages/community/[slug].tsx]: made the changes please refer to the commit- [817c2aa](https://github.com/keploy/blog-website/pull/31/commits/817c2aac2993b1939da6a7049aaf723305525bde)

**Comment by its-kunal** [comment]
Hey @Hermione2408, I have made the changes as asked.

**Comment by Hermione2408** [comment]
Hey @its-kunal 
[Screencast from 01-04-24 01:49:47 PM IST.webm](https://github.com/keploy/blog-website/assets/78696767/cc132fad-7ccb-4168-9dc7-e9656dfec785)
The current hover effect on the Table of Contents (TOC) expands the text, causing it to disrupts the TOC layout. I suggest a subtler hover effect that changes the text color without altering the text size, in line with the Keploy theme (orange).

Additionally, I was thinking to add a side indicator line next to the TOC item that corresponds to the section currently in view. Would love to hear your thoughts on this!

**Comment by its-kunal** [comment]
Hey @Hermione2408, It seems a great idea we can implement a sidebar also for smaller screens to make page navigation more accessible. Here's the reference - [Link](https://dribbble.com/shots/22028039-Taskline-Sidebar-Navigation).

I make changes for subtler hover effects and open a new issue for the sidebar.

File: components/TableContents.tsx [hunk 4]
```diff
@@ -26,7 +27,7 @@ function TOCItem({ id, title, type, onClick }) {
     <li className={itemClasses} style={{ marginLeft }}>
       <button
         onClick={() => onClick(id)}
-        className="block py-1 w-full rounded-md text-left text-sm"
+        className="block py-1 w-full rounded-md text-left text-sm text-black opacity-75 hover:opacity-100 transition-all duration-150 ease-in-out"
```
Comment (Hermione2408) [inline components/TableContents.tsx:30]: you can add this style here hover:text-orange-500

**Review by Hermione2408 — CHANGES_REQUESTED** [review]
not exactly sidebar but that particular section should get activated maybe with a bar on the side of that section in table of content if possible that is also in large view because in small view we get list view so won't need in that

File: components/TableContents.tsx (hunk 4 above)
Comment (its-kunal) [reply components/TableContents.tsx]: Sure made the changes.

**Comment by its-kunal** [comment]
@Hermione2408, It would be a great help if you could reference a design for the particular section.
I think you're looking for this design - [Link](https://dribbble.com/shots/544717-Tab-Sidebar-UI-Concept)
For now, I have made the changes for the right border for a table of content items refer to the commit - 71a72bd

**Comment by Hermione2408** [comment]
> @Hermione2408, It would be a great help if you could reference a design for the particular section. I think you're looking for this design - [Link](https://dribbble.com/shots/544717-Tab-Sidebar-UI-Concept) For now, I have made the changes for the right border for a table of content items refer to the commit - [71a72bd](https://github.com/keploy/blog-website/commit/71a72bdf98b41be075935740b42fd3fcb5dbeb3d)

Hey I didn't meant this you can revert this commit back -71a72bd ,if required will create an issue for it.

**Comment by its-kunal** [comment]
@Hermione2408, made the changes.

**Comment by Hermione2408** [comment]
> @Hermione2408, made the changes.

LGTM @nehagup

**Review by nehagup — APPROVED** [review]
@its-kunal nice work! Merging this one since there's no blocker. However, IMO, the fade in animations could be replaces with slide in animations with each blog sliding in with a few ms delay. It should be applied to all blog previews, tech, community, tweets etc. 

Please feel free to create another PR for this.

**Pull request merged**